crypto/ec/ecp_s390x_nistp.c

```diff
@@ -0,0 +1,394 @@
+#define S390X_OFF_RES_X(n)		(0 * n)
+#define S390X_OFF_RES_Y(n)		(1 * n)
+#define S390X_OFF_SRC_X(n)		(2 * n)
+#define S390X_OFF_SRC_Y(n)		(3 * n)
+#define S390X_OFF_SCALAR(n)		(4 * n)
+
+/* Offsets of fields in KDSA parameter blocks */
+#define S390X_OFF_R(n)			(0 * n)
+#define S390X_OFF_S(n)			(1 * n)
+#define S390X_OFF_H(n)			(2 * n)
+#define S390X_OFF_K(n)			(3 * n)
+#define S390X_OFF_X(n)			(3 * n)
+#define S390X_OFF_RN(n)			(4 * n)
+#define S390X_OFF_Y(n)			(4 * n)
+
+static int ec_GFp_s390x_nistp_mul(const EC_GROUP *group, EC_POINT *r,
+                                  const BIGNUM *scalar,
+                                  size_t num, const EC_POINT *points[],
+                                  const BIGNUM *scalars[],
+                                  BN_CTX *ctx, unsigned int fc, int len)
+{
+    unsigned char param[S390X_SIZE_PARAM];
+    BIGNUM *x, *y;
+    const EC_POINT *point_ptr = NULL;
+    const BIGNUM *scalar_ptr = NULL;
+    BN_CTX *new_ctx = NULL;
+    int rc = -1;
+
+    if (ctx == NULL) {
+        ctx = new_ctx = BN_CTX_new();
+        if (ctx == NULL)
+            return 0;
+    }
+
+    BN_CTX_start(ctx);
+
+    x = BN_CTX_get(ctx);
+    y = BN_CTX_get(ctx);
+    if (x == NULL || y == NULL) {
+        rc = 0;
+        goto ret;
+    }
+
+    /*
+     * Use PCC for EC keygen and ECDH key derivation:
+     * scalar * generator and scalar * peer public key,
+     * scalar in [0,order).
+     */
+    if ((scalar != NULL && num == 0 && BN_is_negative(scalar) == 0)
+        || (scalar == NULL && num == 1 && BN_is_negative(scalars[0]) == 0)) {
+
+        if (num == 0) {
+            point_ptr = EC_GROUP_get0_generator(group);
+            scalar_ptr = scalar;
+        } else {
+            point_ptr = points[0];
+            scalar_ptr = scalars[0];
+        }
+
+        if (EC_POINT_is_at_infinity(group, point_ptr) == 1
+            || BN_is_zero(scalar_ptr)) {
+            rc = EC_POINT_set_to_infinity(group, r);
+            goto ret;
+        }
+
+        memset(&param, 0, sizeof(param));
+
+        if (group->meth->point_get_affine_coordinates(group, point_ptr,
+                                                      x, y, ctx) != 1
+            || BN_bn2binpad(x, param + S390X_OFF_SRC_X(len), len) == -1
+            || BN_bn2binpad(y, param + S390X_OFF_SRC_Y(len), len) == -1
+            || BN_bn2binpad(scalar_ptr,
+                            param + S390X_OFF_SCALAR(len), len) == -1
+            || s390x_pcc(fc, param) != 0
+            || BN_bin2bn(param + S390X_OFF_RES_X(len), len, x) == NULL
+            || BN_bin2bn(param + S390X_OFF_RES_Y(len), len, y) == NULL
+            || group->meth->point_set_affine_coordinates(group, r,
+                                                         x, y, ctx) != 1)
+            goto ret;
+
+        rc = 1;
+    }
+
+ret:
+    /* Otherwise use default. */
+    if (rc == -1)
+        rc = ec_wNAF_mul(group, r, scalar, num, points, scalars, ctx);
```

Comment:
Is this line supposed to run when EC_POINT_is_at_infinity(group, point_ptr) == 1 ?

Reply:
No, Its only supposed to run if iunitial rc=-1 was not touched (man pages says EC_POINT_is_at_inf returns 0 resp 1).

Reply:
So what I was trying to say is, if the ec_wNAF_mul() should only happen when the get_affine/set_affine happens and not anywhere else then maybe move it up to line 104?

Reply:
no, every goto ret with rc at -1 means cant use hw support for some reason, use SW. eg any of the BN_bn2binpad might fail due unusual/out of range x/y/scalar inputs and even s390x_pcc can fail for these out of range inputs, since BN_bn2binpad only detects out of range if its at least whole byte more.

The same comment i quoted from in my other question states that those unusual inputs should be processed for robustness only and that sca resistance is not required for those cases.

Reply:
and wNAF  os used if "if-condition" at 74 evals to false, so cant move wNAF to 104.